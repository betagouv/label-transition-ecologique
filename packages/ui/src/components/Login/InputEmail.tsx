import {Ref, forwardRef} from 'react';
import {FieldError} from 'react-hook-form';
import {Field, FieldMessageProps} from '@design-system/Field';
import {Input} from '@design-system/Input';
import {LoginPropsWithState} from './type';

/**
 * Champ de saisi de l'email
 *
 * Affiche le message appropri√© (info ou erreur).
 */
export const InputEmail = forwardRef(
  (
    {
      loginState,
      infoMessage,
    }: {
      loginState: LoginPropsWithState['loginState'];
      infoMessage?: string;
    },
    ref?: Ref<HTMLInputElement>
  ) => {
    const {
      emailProps,
      form: {
        formState: {errors},
      },
    } = loginState;
    const messageProps: FieldMessageProps = getMessageProps(
      infoMessage,
      errors?.email
    );

    return (
      <Field title="Email de connexion" {...messageProps} htmlFor="email">
        <Input
          className="md:col-span-2"
          id="email"
          type="text"
          autoComplete="on"
          state={errors?.email ? 'error' : undefined}
          ref={ref}
          {...emailProps}
        />
      </Field>
    );
  }
);
const getMessageProps = (
  info: string | undefined,
  error: FieldError | undefined
) => {
  if (error) {
    return {message: error.message, state: 'error'} as FieldMessageProps;
  }
  if (info) {
    return {message: info, state: 'info'} as FieldMessageProps;
  }
  return {};
};
