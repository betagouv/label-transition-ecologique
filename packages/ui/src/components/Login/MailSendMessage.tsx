import MailSend from '../../assets/MailSend';

/** Affiche un message suite Ã  l'envoi du mail de connexion ou de
 * rÃ©initialisation du mot de passe */
export const MailSendMessage = ({
  message1,
  message2,
}: {
  message1: string;
  message2: string;
}) => {
  return (
    <div className="flex flex-col items-center">
      <MailSend />
      <p className="font-bold text-primary-9 text-center">
        {message1}{' '}
        <span className="font-extrabold text-primary-6">{message2}</span>
      </p>
    </div>
  );
};
