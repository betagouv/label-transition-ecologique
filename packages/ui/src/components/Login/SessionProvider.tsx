import {ReactNode, createContext, useEffect, useState} from 'react';
import {Session, SupabaseClient} from '@supabase/supabase-js';
import {clearAuthTokens, setAuthTokens} from './authTokens';

const SessionContext = createContext<Session | null>(null);

type SessionProviderProps = {
  children: ReactNode;
  /** Nom du domaine principal auquel les cookies de session sont rattachés */
  domain: string;
  /** Client sur lequel écouter les événements d'authentification */
  supabase: SupabaseClient;
};

/**
 * Ecoute les événements d'authentification et fourni les informations de session.
 */
export const SessionProvider = ({
  children,
  domain,
  supabase,
}: SessionProviderProps) => {
  const [currentSession, setCurrentSession] = useState<Session | null>(null);

  // enregistre l'écouteur d'événements
  useEffect(() => {
    const {
      data: {subscription},
    } = supabase.auth.onAuthStateChange((event, session) => {
      // efface la session et les cookies à la déconnexion
      if (event === 'SIGNED_OUT') {
        setCurrentSession(null);
        clearAuthTokens(domain);
      }
      // enregistre la session et les cookies lors de la connexion
      else if (
        session &&
        (event === 'SIGNED_IN' || event === 'TOKEN_REFRESHED')
      ) {
        setCurrentSession(session);
        setAuthTokens(session, domain);
      }
    });

    // supprime l'écouteur qd le compo est démonté
    return () => {
      subscription.unsubscribe();
    };
  }, [supabase.auth, domain]);

  // expose la session
  return (
    <SessionContext.Provider value={currentSession}>
      {children}
    </SessionContext.Provider>
  );
};
