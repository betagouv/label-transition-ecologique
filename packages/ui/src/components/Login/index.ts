export * from './Login';
export * from './LoginModal';
export * from './SessionProvider';
export * from './authTokens';
export * from './type';