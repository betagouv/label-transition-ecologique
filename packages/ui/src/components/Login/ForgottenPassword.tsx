import {Button} from '@design-system/Button';
import {
  ModalFooter,
  ModalFooterOKCancel,
  ModalFooterSection,
} from '@design-system/Modal';
import {LoginPropsWithState} from './type';
import {InputEmail} from './InputEmail';
import {FieldMessage} from '@design-system/Field';

/** Demande de réinitialisation du mot de passe */
export const ForgottenPassword = (props: LoginPropsWithState) => {
  const {error, isLoading, loginState, setView, onCancel} = props;
  const {
    contactSupport,
    onSubmitForm,
    form: {
      formState: {isValid},
    },
  } = loginState;

  return (
    <>
      <InputEmail
        loginState={loginState}
        infoMessage="Vous devez être en mesure d'accéder à cette boîte mail pour ouvrir l'e-mail avec le nouveau mot de passe sinon contacter le support."
      />
      <Button
        variant="underlined"
        onClick={e => {
          e.preventDefault();
          contactSupport();
        }}
      >
        Contactez le support !
      </Button>
      {error && (
        <FieldMessage messageClassName="mt-4" state="error" message={error} />
      )}
      <ModalFooter variant="space">
        <Button
          variant="outlined"
          icon="arrow-left-line"
          iconPosition="left"
          onClick={() => setView('par_mdp')}
        >
          Se connecter
        </Button>
        <ModalFooterSection>
          <ModalFooterOKCancel
            btnOKProps={{
              onClick: onSubmitForm,
              disabled: !isValid || isLoading,
            }}
            btnCancelProps={{onClick: onCancel}}
          />
        </ModalFooterSection>
      </ModalFooter>
    </>
  );
};
