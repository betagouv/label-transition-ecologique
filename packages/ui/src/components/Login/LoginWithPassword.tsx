import {Field} from '@design-system/Field';
import {Input} from '@design-system/Input';
import {Button} from '@design-system/Button';
import {FormSectionGrid} from '@design-system/FormSection';
import {InputEmail} from './InputEmail';
import {LoginPropsWithState} from './type';

/**
 * Affiche le formulaire de connexion avec mot de passe
 */
export const LoginWithPassword = ({
  loginState,
  setView,
}: LoginPropsWithState) => {
  const {passwordProps} = loginState;

  return (
    <FormSectionGrid>
      <InputEmail loginState={loginState} />
      <Field
        className="mb-6 md:col-span-2"
        title="Mot de passe"
        htmlFor="password"
      >
        <Input type="password" {...passwordProps} id="password" />
      </Field>
      <Button
        data-test="forgotten-pwd"
        variant="underlined"
        onClick={e => {
          e.preventDefault();
          setView('mdp_oublie');
        }}
      >
        Mot de passe oubli√© ?
      </Button>
    </FormSectionGrid>
  );
};
