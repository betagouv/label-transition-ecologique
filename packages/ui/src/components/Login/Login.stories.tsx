import {Meta, StoryObj} from '@storybook/react';
import {action} from '@storybook/addon-actions';
import {Login} from './Login';

const meta: Meta<typeof Login> = {
  component: Login,
  args: {
    onSubmit: action('onSubmit'),
  },
};

export default meta;

type Story = StoryObj<typeof Login>;

export const Default: Story = {
  args: {},
};

export const AvecMotDePasse: Story = {
  args: {defaultView: 'par_mdp'},
};

export const MotDePasseOublie: Story = {
  args: {defaultView: 'mdp_oublie'},
};

export const MsgLienEnvoye: Story = {
  args: {defaultView: 'msg_lien_envoye'},
};

export const MsgInitMdp: Story = {
  args: {defaultView: 'msg_init_mdp'},
};
