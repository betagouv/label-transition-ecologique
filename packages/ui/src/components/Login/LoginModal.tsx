import {useState} from 'react';
import {Modal} from '@design-system/Modal';
import {LoginProps} from './type';
import {Login} from './Login';

/**
 * Encapsule le panneau d'authentification dans une modale
 */
export const LoginModal = (props: LoginProps) => {
  const [isOpen, setIsOpen] = useState(true);
  const {view, onCancel} = props;
  const size =
    view === 'msg_init_mdp' || view === 'msg_lien_envoye' ? 'md' : 'lg';
  const onClose = () => {
    setIsOpen(false);
    onCancel();
  };
  return (
    <Modal
      dataTest="SignInPage"
      backdropBlur
      size={size}
      openState={{isOpen, setIsOpen}}
      onClose={onClose}
      render={() => <Login {...props} onCancel={() => onClose()} />}
    />
  );
};
