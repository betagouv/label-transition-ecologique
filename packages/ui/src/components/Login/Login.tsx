import {Tab, Tabs} from '@design-system/Tabs';
import {ModalFooterOKCancel} from '@design-system/Modal';
import {LoginProps} from './type';
import {useLoginState} from './useLoginState';
import {ForgottenPassword} from './ForgottenPassword';
import {InputEmail} from './InputEmail';
import {MailSendMessage} from './MailSendMessage';
import {LoginWithPassword} from './LoginWithPassword';

/**
 * Affiche le panneau d'authentification et le formulaire "mot de passe oublié"
 */
export const Login = (props: LoginProps) => {
  const {onCancel, isLoading, view, setView} = props;
  const loginState = useLoginState(props);
  const {
    onSubmitForm,
    form: {
      formState: {isValid},
    },
  } = loginState;

  if (view === 'msg_lien_envoye') {
    return (
      <MailSendMessage
        message1="Pour vous connecter, veuillez consultez votre boite mail et"
        message2="cliquer sur le lien sécurisé reçu !"
      />
    );
  }

  if (view === 'msg_init_mdp') {
    return (
      <MailSendMessage
        message1="Veuillez consultez votre boite mail pour"
        message2="réinitialiser votre mot de passe."
      />
    );
  }

  const title =
    view === 'mdp_oublie' ? 'Mot de passe oublié ? ' : 'Se connecter';

  return (
    <form className="flex flex-col gap-4" onSubmit={onSubmitForm}>
      <h3 className="self-center">{title}</h3>
      {view === 'mdp_oublie' ? (
        <ForgottenPassword {...props} loginState={loginState} />
      ) : (
        <>
          <Tabs
            className="justify-center"
            defaultActiveTab={view === 'par_lien' ? 0 : 1}
            onChange={activeTab => {
              setView(activeTab === 0 ? 'par_lien' : 'par_mdp');
            }}
          >
            <Tab label="Connexion sans mot de passe">
              <InputEmail
                loginState={loginState}
                infoMessage="La connexion se fait par lien sécurisé envoyé sur votre boite mail."
              />
            </Tab>
            <Tab label="Connexion avec mot de passe">
              <LoginWithPassword {...props} loginState={loginState} />
            </Tab>
          </Tabs>
          <ModalFooterOKCancel
            btnOKProps={{
              onClick: onSubmitForm,
              disabled: !isValid || isLoading,
            }}
            btnCancelProps={{onClick: onCancel}}
          />
        </>
      )}
    </form>
  );
};
