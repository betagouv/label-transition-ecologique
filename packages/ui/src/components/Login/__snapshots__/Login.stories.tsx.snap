// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Login AvecMotDePasse smoke-test 1`] = `""`;

exports[`components/Login Default smoke-test 1`] = `""`;

exports[`components/Login MotDePasseOublie smoke-test 1`] = `""`;

exports[`components/Login MsgInitMdp smoke-test 1`] = `""`;

exports[`components/Login MsgLienEnvoye smoke-test 1`] = `""`;
